gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXManagerImpl.java

```diff
@@ -639,13 +646,30 @@ public final void setTXState(final TXStateInterface tx) {
       this.waitForPhase2Commit = false;
     }
 
+    public final void setSnapshotTXState(final TXStateInterface tx) {
+      this.snapshotTXState = tx;
+      this.commitRecipients = null;
+      this.waitForPhase2Commit = false;
+    }
+
     public final void clearTXState() {
       this.txState = null;
       this.commitRecipients = null;
       this.waitForPhase2Commit = false;
       this.msgUsesTXProxy = false;
     }
 
+    public final void clearTXStateAll() {
+      clearTXState();
+      this.snapshotTXState = null;
+    }
+
+    /** clear the TXContext and remove from global list when thread closes */
+    public final void threadClose() {
+      clearTXStateAll();
```

Comment:
Wondering if it should just be cleared or rolledback?

Reply:
Had thought about it but was a bit circumspect if it could mean deadlocks or other issues given that other pools would also have been shutdown.

Reply:
@suranjan Ok, pushed a change to rollback here if cache is not closing -- please check